### PR TITLE
Feat(codesnippet): add experimental component Code Snippet, Closes #1350

### LINKED ISSUE #1350: Experimental Component - Code Snippet - Dev
**Development**
A developer can discover and install the experimental component by following accessible documentation.

- [ ] Add to carbon-components
- [ ] Update website (add to the experimental section and update component status page)

### PR DESCRIPTION
**Pull request by jnm2377**
Closes #1350 

#### Changelog

**New**
- Experimental Code Snippet
- Add styling for copy feedback (was not using Copy Button styling in experimental component)

**Removed**
- Line 44 in `code-snippet.js` that added expand class on init (multi line code would init expanded)

#### Testing / Reviewing
- Ensure multi line code snippet is _not_ expanded on init in both Experimental/Standard
- Ensure copy feedback is the correct color in both Experimental/Standard
- Staging link: http://carbon-dev-environment-cs-xx-noisy-meerkat.mybluemix.net/?nav=code-snippet

### CODE REVIEW

File: src/components/code-snippet/_code-snippet.scss [hunk 1]
```diff
@@ -294,3 +294,326 @@
...
+    background: #3d3d3d;
+  }
+
+  // Show more / less button
+  .#{$prefix}--snippet-btn--expand {
+    position: absolute;
+    right: 0;
+    bottom: 0;
+    padding: rem(8px);
+    padding-left: rem(16px);
```
Comment (aledavila) [inline src/components/code-snippet/_code-snippet.scss]: Do we want to continue using spacing tokens

File: src/components/code-snippet/_code-snippet.scss [hunk 2]
```diff
@@ -294,3 +294,326 @@
...
+    z-index: z('overlay');
+    font-weight: 400;
+    left: inherit;
+    top: 0.75rem;
+    right: 1.25rem;
+  }
+
+  .#{$prefix}--btn--copy__feedback:before,
+  .#{$prefix}--btn--copy__feedback:after {
+    background: #3d3d3d;
```
Comment (aledavila) [inline src/components/code-snippet/_code-snippet.scss]: Change color to the color token. 

Its `$ibm-colors__gray--80` but you might need to add a token for the snippet in theme

File: src/components/code-snippet/_code-snippet.scss [hunk 3]
```diff
@@ -294,3 +294,326 @@
...
+    padding: 0 $spacing-3xl 0 $spacing-md;
+  }
+
+  .#{$prefix}--snippet--single .#{$prefix}--snippet-container {
+    display: flex;
+    align-items: center;
+    height: rem(32px);
+    overflow-x: scroll;
+    position: relative;
+    padding: 0 0 16px 0;
```
Comment (aledavila) [inline src/components/code-snippet/_code-snippet.scss]: Spacing tokens

**Review by aledavila — COMMENTED** [review]
looks good just left a few comments

**Comment by dakahn** [comment]
Aside from Ale's requested changes. It looks like solid work to me. 🥇

**Review by dakahn — APPROVED** [review]

File: src/components/code-snippet/_code-snippet.scss [hunk 4]
```diff
@@ -294,3 +294,326 @@
...
+    position: absolute;
+    right: 0;
+    bottom: 0;
+    padding: rem(8px);
+    padding-left: rem(16px);
+    color: $text-01;
+  }
+
+  .#{$prefix}--snippet-btn--expand--hide.#{$prefix}--snippet-btn--expand {
+    //if you add a space between classes, button disappears?
```
Comment (asudoh) [inline src/components/code-snippet/_code-snippet.scss]: Did you have a question here...?

File: src/components/code-snippet/_code-snippet.scss [hunk 5]
```diff
@@ -294,3 +294,326 @@
...
+      width: 95%;
+    }
+  }
+
+  .#{$prefix}--snippet--single.#{$prefix}--skeleton .#{$prefix}--snippet-container {
+    padding-bottom: 0;
+  }
+
+  /*
+    Deprecated class names will be removed in v10.x
```
Comment (asudoh) [inline src/components/code-snippet/_code-snippet.scss]: From reading this comment - Do we need `--terminal` things in experimental?

File: src/components/code-snippet/code-snippet.js [hunk 6]
```diff
@@ -41,7 +41,6 @@ class CodeSnippet extends mixin(createComponent, initComponentBySearch, handles)
 
     if (this.element.offsetHeight < this.options.minHeight) {
       this.element.classList.add(this.options.classHideExpand);
-      this.element.classList.add(this.options.classExpanded);
```
Comment (asudoh) [inline src/components/code-snippet/code-snippet.js:44]: Is this change intentional? Also, does the stable mode still work?

**Review by asudoh — COMMENTED** [review]
Thanks @jnm2377 for making this great progress 🎉!

File: src/components/code-snippet/_code-snippet.scss [hunk 7]
```diff
@@ -294,3 +294,326 @@
...
+    cursor: pointer;
+
+    &:hover {
+      background-color: $ui-03;
+    }
+
+    &:focus {
+      outline: none;
+      border: 2px solid $brand-01;
+    }
```
Comment (alisonjoseph) [inline src/components/code-snippet/_code-snippet.scss:324]: Can we use the focus mixin here, or no because of the rounded corners? It's causing the spacing to jump so if we can't use `outline` we may need to use a transparent border as default to avoid the shift. 

I think we can remove lines 322/333 and add ` @include focus-outline('blurred');`

File: src/components/code-snippet/_code-snippet.scss (hunk 5 above)
Comment (alisonjoseph) [reply src/components/code-snippet/_code-snippet.scss]: Agreed

 `--terminal` and `--code` can be removed from experimental

**Comment by alisonjoseph** [comment]
This looks great other than a couple little updates needed! Excited to get this implemented in the new website design 🎉

File: src/components/code-snippet/_code-snippet.scss (hunk 2 above)
Comment (jnm2377) [reply src/components/code-snippet/_code-snippet.scss]: Shoot, I totally meant to change that. Thanks for catching that @aledavila !

File: src/components/code-snippet/_code-snippet.scss (hunk 7 above)
Comment (jnm2377) [reply src/components/code-snippet/_code-snippet.scss:324]: You're right, @alisonjoseph. Thanks for that catch! I realized the transparent border width was different than the focus border width, that's why it was jumping. I fixed it. :)

File: src/components/code-snippet/code-snippet.js (hunk 6 above)
Comment (jnm2377) [reply src/components/code-snippet/code-snippet.js:44]: @asudoh This change was intentional. When I added the styling for the experimental component, the multi line code snippet would initiate already expanded, but the button would still read `Show More`. Then when I clicked it, it would fix to `Show Less`, and it would work perfectly fine after that. When I disabled the feature-flag to check if the standard multi line code snippet had that same issue, it didn't, it would initiate normally. When I removed that line of code, the experimental component would initiate properly and so would standard. I'm not sure what stable mode is.

File: src/components/code-snippet/_code-snippet.scss (hunk 3 above)
Comment (jnm2377) [reply src/components/code-snippet/_code-snippet.scss]: 👍

File: src/components/code-snippet/_code-snippet.scss (hunk 4 above)
Comment (jnm2377) [reply src/components/code-snippet/_code-snippet.scss]: Not really. It was more of a comment for myself. I went ahead and deleted it. :)

**Review by joshblack — APPROVED** [review]

**Review by aagonzales — CHANGES_REQUESTED** [review]
Sorry we've made a couple of **design updates**: 

- [x] The design language team decided to make 40px the main component height and not 32px. So can we actually change the height on the `single line` snippet to be 40px? 
   - And then the copy button would be in a 40x40px box
   - As an aside, I think in the future we'll want small (32px), default (40px) and large (48 or 64) sizes.
image
image

- [x] Then for `multi-line` snippet the copy and show more stay the same at 32x32 but get indented by 8px on top and bottom from the container edge
image
image

- [x] `copy` icon should be Gray-100 or new token `icon-01`. We changed that after the color audit or its now wrong in the design file, that's my bad.

**Bugs:**
- [ ] the `show more/ show less` function lags a little after click. Its worse in firefox than it is in chrome.
- [x] the inline snippet is pushing content when you click on it. Ideally it wouldn't affect the content. 
nov-16-2018 11-48-07
- [x] the chevron in the `show more/ show less` button looks off center and a little too low.

**Comment by jnm2377** [comment]
@IBM/carbon-developers is the `inline snippet ` still jumping for y'all? It's not jumping for me anymore both locally or in the staging link, but designers said it was jumping for them in Chrome.

**Comment by jnm2377** [comment]
@aagonzales ready for review again. Here's an updated [staging link](http://carbon-dev-environment-cs-xx-noisy-meerkat.mybluemix.net/?nav=code-snippet).

**Review by aagonzales — APPROVED** [review]
Looks good!! Thanks for making those updates! 👍 ✅
